src/transformers/generation_logits_process.py

```diff
@@ -372,3 +373,25 @@ def _set_scores_to_inf_for_banned_tokens(self, scores: torch.Tensor, banned_toke
         )
         scores = scores.masked_fill(banned_mask, -float("inf"))
         return scores
+
+
+class PrefixConstrainedLogitsProcessor(LogitsProcessor):
+    r"""
+    :class:`transformers.LogitsProcessor` .
+
+    Args:
+        prefix_allowed_tokens_fn (:obj:`Callable`):
+
+    """
+
+    def __init__(self, prefix_allowed_tokens_fn: Callable, num_beams: int):
+        self._prefix_allowed_tokens_fn = prefix_allowed_tokens_fn
+        self._num_beams = num_beams
+
+    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
+        mask = torch.full_like(scores, -math.inf)
+        for batch_id, beam_sent in enumerate(input_ids.view(-1, self._num_beams, input_ids.shape[-1])):
```

Comment:
In a future PR we could probably speed this up by just using `torch.Tensor` operations and not Python loops. Python loops really slow down the computation on GPU apparently (see: https://github.com/huggingface/transformers/pull/6064). But we can do this in a future PR as well

Reply:
I wanted to keep the same signature as in fairseq as if someone has already implemented one it can use the same ;)